Zend/zend_enum.c

```diff
@@ -0,0 +1,350 @@
+		if (Z_TYPE_P(zv) == IS_CONSTANT_AST) {
+			zval_update_constant_ex(zv, c->ce);
+			if (UNEXPECTED(EG(exception) != NULL)) {
+				return;
+			}
+		}
+		if (ce->enum_scalar_type == IS_UNDEF) {
+			zend_hash_next_index_insert(Z_ARRVAL_P(return_value), zv);
+		} else {
+			zend_object *zobj = Z_OBJ_P(zv);
+			zend_string *value_string = zend_string_init("value", strlen("value"), 0);
+			zval tmp;
+			zval *value_zv = zend_read_property_ex(ce, zobj, value_string, 0, &tmp);
+			zend_string_free(value_string);
+
+			if (ce->enum_scalar_type == IS_LONG) {
+				zend_hash_index_update(Z_ARRVAL_P(return_value), Z_LVAL_P(value_zv), zv);
+			} else {
+				ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+				zend_symtable_update(Z_ARRVAL_P(return_value), Z_STR_P(value_zv), zv);
+			}
+		}
+	} ZEND_HASH_FOREACH_END();
+}
+
+static ZEND_NAMED_FUNCTION(zend_enum_from_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_string *string_key;
+	zend_long long_key;
+
+	zval *case_name_zv;
+	if (ce->enum_scalar_type == IS_LONG) {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &long_key) == FAILURE) {
+			RETURN_THROWS();
+		}
+
+		case_name_zv = zend_hash_index_find(ce->enum_scalar_table, long_key);
+	} else {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "S", &string_key) == FAILURE) {
+			RETURN_THROWS();
+		}
+
+		ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+		case_name_zv = zend_hash_find_ex(ce->enum_scalar_table, string_key, 1);
+	}
+
+	if (case_name_zv == NULL) {
+		if (ce->enum_scalar_type == IS_LONG) {
+			zend_value_error("%d is not a valid scalar value for enum \"%s\"", (int) long_key, ZSTR_VAL(ce->name));
+		} else {
+			ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+			zend_value_error("\"%s\" is not a valid scalar value for enum \"%s\"", ZSTR_VAL(string_key), ZSTR_VAL(ce->name));
+		}
+		RETURN_THROWS();
+	}
+
+	ZEND_ASSERT(Z_TYPE_P(case_name_zv) == IS_STRING);
+	zval *case_const_zv = zend_hash_find_ex(&ce->constants_table, Z_STR_P(case_name_zv), 0);
+	ZEND_ASSERT(case_const_zv != NULL);
+	zend_class_constant *c = Z_PTR_P(case_const_zv);
+	zval *case_zv = &c->value;
+	if (Z_TYPE_P(case_zv) == IS_CONSTANT_AST) {
+		zval_update_constant_ex(case_zv, c->ce);
+		if (UNEXPECTED(EG(exception) != NULL)) {
+			RETURN_THROWS();
+		}
+	}
+
+	ZVAL_COPY(return_value, case_zv);
+}
+
+static ZEND_NAMED_FUNCTION(zend_enum_has_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_string *string_key;
+	zend_long long_key;
+
+	zval *case_name_zv;
+	if (ce->enum_scalar_type == IS_LONG) {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &long_key) == FAILURE) {
+			RETURN_THROWS();
+		}
+
+		case_name_zv = zend_hash_index_find(ce->enum_scalar_table, long_key);
+	} else {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "S", &string_key) == FAILURE) {
+			RETURN_THROWS();
+		}
+
+		ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+		case_name_zv = zend_hash_find_ex(ce->enum_scalar_table, string_key, 1);
```

Comment:
```
static zend_always_inline zval *zend_hash_find_ex(const HashTable *ht, zend_string *key, zend_bool known_hash)
{
	if (known_hash) {
		return _zend_hash_find_known_hash(ht, key);
	} else {
		return zend_hash_find(ht, key);
	}
}
```

I don't think the hash of string_key is known - `zend_hash_find` would make more sense to me? e.g. `SomeEnum::has(strtolower('SOMETHING'))` would be a string generated at runtime where the hash probably wasn't computed yet
```suggestion
		case_name_zv = zend_hash_find(ce->enum_scalar_table, string_key);
```

Reply:
I see the bug I expected

```php
php > enum Test2: string { case lower='lower'; } var_export(Test2::has(strtolower('LOWER')));
false
php > var_export(Test2::has('lower'));
true
```

Reply:
Thanks, tested and fixed